modules/sdk/src/main/scala/org/tessellation/sdk/domain/cluster/programs/L0PeerDiscovery.scala

```diff
@@ -28,17 +31,44 @@ object L0PeerDiscovery {
   }
 }
 
-sealed abstract class L0PeerDiscovery[F[_]: Sync] private (
+sealed abstract class L0PeerDiscovery[F[_]: Sync: Random] private (
   l0ClusterClient: L0ClusterClient[F],
   l0ClusterStorage: L0ClusterStorage[F]
 ) {
 
   val logger = Slf4jLogger.getLogger[F]
 
-  def discoverFrom(peer: P2PContext): F[Unit] =
+  def discover(lastFacilitators: NonEmptySet[PeerId]): F[Unit] =
+    l0ClusterStorage.getPeers
+      .map(_.map(_.id).intersect(lastFacilitators))
+      .map(_.toList)
+      .flatMap(Random[F].shuffleList)
+      .map(_.head)
+      .flatMap(l0ClusterStorage.getPeer)
+      .map(_.map(p => P2PContext(p.ip, p.port, p.id)))
+      .flatMap(_.fold(Applicative[F].unit) { p =>
+        getPeersFrom(p).flatMap(l0ClusterStorage.setPeers)
+      })
+      .handleErrorWith { error =>
+        logger.warn(error)(s"An error occured during L0 peer discovery")
+      }
+
+  private def getPeersFrom(peer: P2PContext): F[NonEmptySet[L0Peer]] =
     l0ClusterClient.getPeers
       .run(peer)
       .map(_.map(L0Peer.fromPeer))
+      .flatMap { s =>
+        NonEmptySet
+          .fromSet(SortedSet.from(s))
+          .fold {
+            (new Throwable("Unexpected state - no peers found but at least one should be available")).raiseError[F, NonEmptySet[L0Peer]]
+          }(Applicative[F].pure)
+      }
+
+  def discoverFrom(peer: P2PContext): F[Unit] =
+    getPeersFrom(peer)
+      .map(_.toSortedSet)
+      .map(_.toSet)
```

Comment:
`addPeers` will not accept the sortedSet?

Reply:
![image](https://user-images.githubusercontent.com/4959931/181641703-bde9ae78-8790-42d5-a050-e2eb7e23718f.png)

Reply:
I mean isn't sortedSet a set?